ee/desktop/runner/runner_linux.go

```diff
@@ -104,31 +112,98 @@ func (r *DesktopUsersProcessesRunner) userEnvVars(uid string) map[string]string
 		}
 
 		sessionType := strings.Trim(string(typeOutput), "\n")
-		switch sessionType {
-		case "x11":
-			// We need to set DISPLAY
-			xDisplayOutput, err := exec.CommandContext(ctx, "loginctl", "show-session", session, "--value", "--property=Display").Output()
-			if err != nil {
-				level.Debug(r.logger).Log(
-					"msg", "could not get Display from user session",
-					"uid", uid,
-					"err", err,
-				)
-				continue
-			}
+		if sessionType == "x11" {
+			envVars["DISPLAY"] = r.displayFromX11(session, ctx)
+		} else if sessionType == "wayland" {
+			// For opening links with x-www-browser, we only need DISPLAY.
+			envVars["DISPLAY"] = r.displayFromXwayland(int32(uidInt), ctx)
+
+			// For opening links with xdg-open, we need XDG_DATA_DIRS so that xdg-open can find the mimetype configuration
+			// files to figure out what application to launch.
+
+			// We take the default value according to https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html,
+			// but also include the snapd directory due to an issue on Ubuntu 22.04 where the default
+			// /usr/share/applications/mimeinfo.cache does not contain any applications installed via snap.
+			envVars["XDG_DATA_DIRS"] = "/usr/local/share/:/usr/share/:/var/lib/snapd/desktop"
+		} else {
+			// Not a graphical session
+			continue
+		}
+	}
+
+	return envVars
+}
+
+func (r *DesktopUsersProcessesRunner) displayFromX11(session string, ctx context.Context) string {
+	// We can read $DISPLAY from the session properties
+	xDisplayOutput, err := exec.CommandContext(ctx, "loginctl", "show-session", session, "--value", "--property=Display").Output()
+	if err != nil {
+		level.Debug(r.logger).Log(
+			"msg", "could not get Display from user session",
+			"err", err,
+		)
+		return defaultDisplay
+	}
+
+	return strings.Trim(string(xDisplayOutput), "\n")
+}
+
+func (r *DesktopUsersProcessesRunner) displayFromXwayland(uid int32, ctx context.Context) string {
```

Comment:
`ctx` should be the first argument. (I think, this one is not a nit)

Reply:
Updated all method signatures here so ctx is first ✅

Reply:
Thank you